src/include/sof/audio/ipcgtw_copier.h

```diff
@@ -34,6 +35,66 @@ struct ipcgtw_data {
 	uint32_t buf_size;
 };
 
+/**< IPC header format for IPC gateway messages */
+struct ipc4_ipcgtw_cmd {
+	union  {
+		uint32_t dat;
+
+		struct {
+			/**< Command, see below */
+			uint32_t cmd : 24;
+
+			/**< One of Global::Type */
+			uint32_t type : 5;
+
+			/**< Msg::MSG_REQUEST */
+			uint32_t rsp : 1;
+
+			/**< Msg::FW_GEN_MSG */
+			uint32_t msg_tgt : 1;
+
+			uint32_t _reserved_0 : 1;
+		} r;
+	} primary;
+	union  {
+		uint32_t dat;
+		struct {
+			uint32_t data_size : 30;
+			uint32_t _reserved_0 : 2;
+		} r;
+	} extension;
+} __packed __aligned(4);
+
+/**< Values of ipc4_ipcgtw_cmd::primary.r.cmd */
+enum {
+	IPC4_IPCGWCMD_GET_DATA = 1,
+	IPC4_IPCGWCMD_SET_DATA = 2,
+	IPC4_IPCGWCMD_FLUSH_DATA = 3
+};
+
+/* Incoming IPC gateway message */
+struct ipc4_ipc_gateway_cmd_data {
+	/* node_id of the target gateway */
+	union ipc4_connector_node_id node_id;
+	/* Payload (actual size is in the header extension.r.data_size) */
+	uint8_t payload[];
+};
```

Comment:
if this structure is transmitted from memory window, we need to use "__packed __aligned(4);", also include "struct ipc4_ipc_gateway_cmd_data_reply"

Reply:
these structure directly copied from ipcgtw.h, no other changes:
in = (const struct ipc4_ipc_gateway_cmd_data *)MAILBOX_HOSTBOX_BASE;
out = (struct ipc4_ipc_gateway_cmd_data_reply *)reply_payload;

itself does not allocate memory, it just get memory from driver.

Reply:
thanks, @RanderWang @ranj063 , I added with a new patch, a practice is compiler will take care of alignment when structure memory allocation, but may not all compiler can cover this.